### ISSUE TITLE
searchlight without cv

### ISSUE THREAD

**Issue by SylvainTakerkart** [nilearn/nilearn#909]
Hello!

I'm trying to use the searchlight object, but I need to have access to the scores_ maps for each fold of the cv... Starting from the haxby example, I tried to modify the code as follows by removing the cv argument and the looping over the cv object (scikit-learn style):

```python
for train_index, test_index in cv:
    # define train and test datasets
    fmri_img_train = index_img(fmri_img, train_index)
    fmri_img_test = index_img(fmri_img, test_index)
    y_train = y[train_index]
    y_test = y[test_index]
    searchlight = nilearn.decoding.SearchLight(
        mask_img,
        process_mask_img=process_mask_img,
        radius=5.6, n_jobs=n_jobs,verbose=1)
    searchlight.fit(fmri_img_train, y_train)
    searchlight.predict(fmri_img_test)
```
but there's no 'predict' method... then I read the doc of the searchlight class, and it says that if no cv is given, it does a 3-fold cv... so I'm stuck now!

Any clue on how to do this?

Thanks,

Sylvain

**Comment by lesteve** [comment]
Have you tried, passing `cv=cv` into the SearchLight object?

```python

searchlight = nilearn.decoding.SearchLight(
        mask_img,
        process_mask_img=process_mask_img,
        radius=5.6, n_jobs=n_jobs,verbose=1,
        cv=cv)
```

**Comment by GaelVaroquaux** [comment]
SearchLight cannot do "predict": the search light method is not a
predictive model. This is a fundemental limitation of the method, not of
our implementation.

**Comment by SylvainTakerkart** [comment]
Hey, thanks for your feedback, but I think my question was not clear... Let me try to rephrase...

I'm trying to do a searchlight analysis where I would have access to the classification accuracy maps of each fold of the cv that I pass with cv=cv... Any suggestion on how to do this? (from what I understand, the searchlight.scores_ map is the average of the maps I'm looking for)

Sylvain

**Comment by GaelVaroquaux** [comment]
> I'm trying to do a searchlight analysis where I would have access to the
> classification accuracy maps of each fold of the cv that I pass with cv=cv...
> Any suggestion on how to do this? (from what I understand, the
> searchlight.scores_ map is the average of the maps I'm looking for)

searchlight.scores_ map is not computed that way.

You want the out-of-bag error for each fold? Then you have to run
multiple searchlights withCV objects that you craft to have only one
fold.

By the way: I don't know what you are trying to do, but it seems a bad
idea :).

**Comment by SylvainTakerkart** [comment]
OK, thanks Gael, and yes, that's what I want! Two more questions then...

1. for info, how is searchlight.scores_ computed?

2. can you guide me (at least to get me started) on how to design CV object with just one fold?

**Comment by eickenberg** [comment]
For 2. : You can use

from sklearn.cross_validation import ShuffleSplit
cv = ShuffleSplit(n_samples, n_iter=1, test_size=.5)

where test_size needs to be set to the fraction of data or the number of
data points that you want in your test fold.

**Comment by SylvainTakerkart** [comment]
thanks, but, once again, I think my question was not properly framed, sorry... I actually want to do a LeaveOneLabelOut cv... so now, I'd need to construct several cvs which would correspond to each fold of my LeaveOneLabelOut cv

**Comment by eickenberg** [comment]
Then maybe: define `cv` however you want, and then

```

results = []
for cv_item in cv:
    one_cv = [cv_item]
    results.append(do_searchlight_analysis(one_cv))


On Tue, Dec 15, 2015 at 2:54 PM, SylvainTakerkart <notifications@github.com>
wrote:

> thanks, but, once again, I think my question was not properly framed,
> sorry... I actually want to do a LeaveOneLabelOut cv... so now, I'd need to
> construct several cvs which would correspond to each fold of my
> LeaveOneLabelOut cv
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/issues/909#issuecomment-164771785>.
>

**Comment by SylvainTakerkart** [comment]
excellent, that does the job, thanks a lot!!

other than that, I would still be interested in knowing how searchlight.scores_ is computed... thanks in advance!

**Comment by banilo** [comment]
One way to view this would be: searchlight performs repeated local function approximation and maps the results to each voxel of the brain (this is why a single predict is ambiguous), whereas other whole-brain classifiers would act on the brain, globally (predict then makes sense).

**Comment by GaelVaroquaux** [comment]
cross_val_score

**Comment by SylvainTakerkart** [comment]
ok, thanks everybody!

(I guess you could also view the searchlight in yet another way for which having a predict method would make sense, but you probably have thought this through more than me ;) )

**Comment by sgagnon** [comment]
How is the searchlight.scores_ map computed, if not by taking the mean across CV folds? It looks like the code is taking the mean across the folds (i.e., the output from `cross_val_scores` - an array of scores (shape = (len(list(cv)),)) of the estimator for each run of the cross validation). 

https://github.com/nilearn/nilearn/blob/master/nilearn/decoding/searchlight.py#L166

**Comment by KamalakerDadi** [comment]
>How is the searchlight.scores_ map computed

It mostly uses cross_val_score and its important parameters are input data (X) and target variables (y) and estimator to fit on the input data.

Please have a look at this link for overview of the searchlight object, particularly 2.4.3.3. Cross validation
http://nilearn.github.io/decoding/searchlight.html

Hope this helps.

**Comment by sgagnon** [comment]
Thanks for the reply. I have a pretty good idea about how the searchlight object works, I'm just specifically focused on how `searchlight.scores_` is computed, given the earlier comment in this thread where "_from what I understand, the searchlight.scores_ map is the average of the maps I'm looking for_" (@SylvainTakerkart) was followed with a direct reply stating that "_searchlight.scores_ map is not computed that way_." My confusion is related to the specific code ~line 166 in `nilearn/decoding/searchlight.py` which says the following:

`par_scores[i] = np.mean(cross_val_score(estimator, X[:, row], y, cv=cv, n_jobs=1, **kwargs))`

Given this code, I'm confused how searchlight.scores_ is *not* simply the mean across CVs, given that `cross_val_score` returns an array of scores of shape = (len(list(cv)),) of the estimator for each run of the cross validation.

**Comment by bthirion** [comment]
@sgagnon yes the score is exactly what you believe it is.
